pkg/workloadapi/workload.proto

```diff
@@ -17,6 +17,55 @@ syntax = "proto3";
 package istio.workload;
 option go_package="pkg/workloadapi";
 
+// Address represents a unique address, keyed by IP address and network.
+// An Address can represent a Service or an individual workload.
+// Address is intended to be able to be looked up on-demand, allowing a workload
+// to answer a question like "what is this IP address", similar to a reverse DNS lookup.
+// Keyed by network/IP address. Example: "default/1.2.3.4".
+//
+// In some cases, we do not know the IP address of a Workload. For instance, we may simply know
+// that there is a workload behind a gateway, and rely on the gateway to handle the rest.
+// In this case, the key format will be "network/resource-uid". The resource can be a Pod, WorkloadEntry, etc.
+// These resources cannot be looked up on-demand.
+//
+// In some cases, we do not know the IP address of a Service. These services cannot be used for matching
+// outbound traffic, as we only have L4 attributes to route based on. However,
+// they can be used for Gateways.
+// In this case, the key format will be "network/hostname".
+// These resources cannot be looked up on-demand.
+message Address {
+  oneof type {
+    // Workload represents an individual workload.
+    // This could be a single Pod, a VM instance, etc.
+    Workload workload = 1;
+    // Service represents a service - a group of workloads that can be accessed together.
+    Service service = 2;
+  }
+}
+
+// Service represents a service - a group of workloads that can be accessed together.
+message Service {
+  // Name represents the name for the service.
+  // For Kubernetes, this is the Service name.
+  string name = 1;
+  // Namespace represents the namespace for the service.
+  string namespace = 2;
+  // Hostname represents the FQDN of the service.
+  // For Kubernetes, this would be <name>.<namespace>.svc.<cluster domain>
+  string hostname = 3;
+  // Address represents the IPv4/IPv6 address for the service.
+  // This should be globally unique.
+  // This should not have a port number.
+  // TODO: Add network as discriminator
+  // TODO: Add dual stack support
+  bytes address = 4;
+  // Ports for the service
+  repeated Port ports = 5;
+  // Optional; if set, the SAN to verify for TLS connections.
+  // Typically, this is not set and per-workload identity is used to verfiy
+  repeated string subject_alt_names = 6;
```

Comment:
This ties to https://github.com/istio/istio/pull/44199#discussion_r1154670647 a bit where we discuss protocol.

We need to decide how to represent services that are not JUST a grouping of Workloads. Example: google.com.

Adding subject_alt_names here is to handle this, really, but its not sufficient. We also need to know the protocol (HBONE or not), and if it even is this type (does it have 0 workloads and we should drop traffic, or its a "pure service" and we should forward as is?).

I think we need to either add more fields to Service OR remove this and have those use cases represented by a Service + a single ~empty Workload.

WDYT?

Sorry I didn't realize this until now...

Reply:
I was thinking the latter (service with a mostly empty workload) when I did https://github.com/istio/ztunnel/pull/486

google.com could be the (External)Service represented as a service and then we could have workload(s) behind it

Reply:
Thinking it through a bit more, I think there 2 are cases

For ztunnel proxying, service+empty workload makes sense. We match on Service IP, send to Workload IP - they can both be the same

For network gw or waypoint, its different. We may have no IP at all, just hostname and do `curl hostname` equivilent. In that case, its ALWAYS hbone though.

So I think it actually works as-is like this, but its just a bit implicit?

Reply:
Maybe for services who are a group of workloads, we can determine protocol based on individual workload that is going to be selected.  for external services who don't have workloads associated, we may need to have protocol info specified in the service xDS resource.

Reply:
I suppose we could want a mix case - match on Service IP (ex: auto-allocated by DNS), but then send to a hostname. But not sure why we want ztunnel to do that, its not adding value there really?

I guess maybe it matters since if we _don't_ do that, then we cannot do auto-allocation or things would be broken (we MUST translate the fake IP)

Reply:
```
We also need to know the protocol (HBONE or not), 
```
Just so I'm clear, the only reason we would need protocol on the service itself is to support "things that are not Workloads that speak HBONE" - correct?

```
and if it even is this type (does it have 0 workloads and we should drop traffic, or its a "pure service" and we should forward as is?).
```

Aren't those sort of the same thing/why do we care to distinguish between them?

Reply:
> I suppose we could want a mix case - match on Service IP (ex: auto-allocated by DNS), but then send to a hostname. But not sure why we want ztunnel to do that, its not adding value there really?

I think we will want to do this for multi network. e.g. user hits a SE in one network (auto allocated) but we still need to use the VIP/attached hostname in the request since the other network can only use the hostname to find another instance of the service. It cannot do anything with the auto allocated VIP

Reply:
just to follow up here, the discussions and new api proposals to resolve this issue live in the latest version of this doc: https://docs.google.com/document/d/1V5wkeBHbLSLMzAMbwFlFZNHdZPyUEspG4lHbnB0UaCg/edit